x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,230 @@
+				slotGroups[i].Count += 1
+				foundGroup = true
+				break
+			}
+		}
+
+		if !foundGroup {
+			newGroup := NewPairingSlotGroup(slots[k])
+			slotGroups = append(slotGroups, *newGroup)
+		}
+	}
+
+	return slotGroups
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// calculates the final pairing score for all slot groups (with strategy)
+// we calculate only the diff between the current and previous slot groups
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight, ok := strategy[reqName]
+		if !ok {
+			return utils.LavaFormatError("req not in strategy", sdkerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "req_name", Value: reqName})
+		}
+
+		for _, score := range scores {
+			// normalize stake so we won't overflow the score result (uint64)
+			newScoreComp := req.Score(*score.Provider, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				if prevReqScoreComp == 0 {
+					return utils.LavaFormatError("previous req score is zero", fmt.Errorf("invalid req score"),
+						utils.Attribute{Key: "req_name", Value: reqName})
+				}
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// given a list of scores, pick a <group-count> providers with a pseudo-random weighted choice
+func PickProviders(ctx sdk.Context, projectIndex string, scores []*PairingScore, groupCount uint64, block uint64, chainID string, epochHash []byte, indexToSkipPtr *map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	scoreSum := sdk.NewUint(0)
+	hashData := make([]byte, 0)
+	for _, providerScore := range scores {
+		scoreSum = scoreSum.Add(sdk.NewUint(providerScore.Score))
+	}
+	if scoreSum.IsZero() {
+		// list is empty
+		return returnedProviders
+	}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	indexToSkip := *indexToSkipPtr
+	for it := 0; it < int(groupCount); it++ {
+		hash := tendermintcrypto.Sha256(hashData) // TODO: we use cheaper algo for speed
+		bigIntNum := new(big.Int).SetBytes(hash)
+		hashAsNumber := sdk.NewUintFromBigInt(bigIntNum)
+		modRes := hashAsNumber.Mod(scoreSum)
+
+		newScoreSum := sdk.NewUint(0)
+		// we loop the servicers list form the end because the list is sorted, biggest is last,
```

Comment:
who sorts it?
(also fix typo)

Reply:
I copied this function from the old pairing mechanism (originally written by Omer), and this is an unwanted relic. Should I just delete the comment or do you think it's necessary to perform sort?

Reply:
depends on the desired functionality. 
it sounds like sorting helps the goal (in terms of scoring strategy). 
it also may help determinism (but does not guarantee order of entries with same score).

Reply:
I sort the scores like expected in the comment. If there are equal scores, I determine their order by the provider's address (alphabetically) to keep determinism. see https://github.com/lavanet/lava/commit/81e1f47bebf82edec3e41942d4f9c428954b4cbd

Reply:
ignore last comment. After talking with Yarom, I removed the sort. see https://github.com/lavanet/lava/commit/d30134e7c735b1c043ac5b7455dca1dff4ad612d